### ISSUE TITLE
spread.banana should be more helpful when user tries to encode unicode

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7663]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7663|
|Type|enhancement|
|Created|2014-10-04 07:17:42Z|
|Branch|https://github.com/twisted/twisted/tree/banana-unsupported-type-exception-7663|

which banana cannot do. This could be a more frequent problem with PY3 where the native string is unicode.

With this patch it raises a specific error message.

Attachments:

* [7663.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/209/209105f4c8d296f5f672054021f7d3d9d21c2176/94597438d9cd6b545c1901678e96b279f732903d.patch) (1977 bytes) - added by wolfgang61 on 2014-10-04 07:44:07Z - 
* [7663b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/209/209105f4c8d296f5f672054021f7d3d9d21c2176/3ac6266a6de91c2b4687f4633dbc82e1361167e2.patch) (3256 bytes) - added by wolfgang61 on 2014-10-04 14:09:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7663 7663
type__enhancement enhancement
reporter__wolfgang61 wolfgang61
priority__normal normal
milestone__ 
branch__branches_banana_unsupported_type_exception_7663 branches/banana-unsupported-type-exception-7663
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__spread spread
time__1412407062064811 1412407062064811
changetime__1412467398635730 1412467398635730
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to wolfgang61|
|-|-|

Thanks.

  1. Use fully qualified names in the news fragment.
  1. I think the test method docstring isn't the most helpful place for that admonishment.  People are more likely to read the banana documentation (though unless they're using banana directly they're probably not likely to read that either).
  1. The existing exception case could be made more informative for all types.  `"Banana cannot send {} objects: {!r}".format(fullyQualifiedName(type(obj)), obj)` would convey the information the new case conveys but also make every other case explicit about the unsupported type too.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wolfgang61 set owner to @exarkun|
|-|-|

New patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [43202]) Branching to 'banana-unsupported-type-exception-7663'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [43203]) Apply 7663b.patch 

refs [#7663](https://github.com/twisted/twisted/issues/7663)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [43205]) Format the documentation in accordance with contemporary style.

refs [#7663](https://github.com/twisted/twisted/issues/7663)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [43206]) The protocol isn't concerned with Python 2 or Python 3.  Just talk about bytes and unicode.

refs [#7663](https://github.com/twisted/twisted/issues/7663)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [43207]) Split up the test with multiple assertions into several tests.

Also adjust the bytes/unicode usage in these tests.
The name of a type is a native string.
The string version of an exception is a native string.

Also be explicit about the expected strings instead of re-computing them the same way
as the implementation computes them.  This means the test will probably fail on Python 3
and will need to be adjusted when Banana is for-real ported to Python 3.

refs [#7663](https://github.com/twisted/twisted/issues/7663)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [43212]) Python 2.6 compatibility

refs [#7663](https://github.com/twisted/twisted/issues/7663)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Thanks.  I made a few more adjustments.  If buildbot turns green I'll go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [43213]) Merge banana-unsupported-type-exception-7663

Author: wolfgang61, exarkun
Reviewer: exarkun
Fixes: [#7663](https://github.com/twisted/twisted/issues/7663)

Change Banana's unsupported type exception message to include the type of the
unsupported object.
